File: utils/cpp/src/Comm/TCP/Port.spec.cpp
#include <catch.hpp>
#include <Comm/TCP/Port.hpp>
#include <Comm/TCP/SingleClientServer.hpp>
#include <thread>
#include <chrono>

#define TEST_ADDRESS "localhost"
#define TEST_PORT 3001
#define PAUSE std::this_thread::sleep_for(std::chrono::milliseconds(10));

/*
  Default timeout takes a long time to throw the exeption slowing down the tests.
*/
TEST_CASE("tcp port throws ConnectionFailure", "[TCP_Port][.timing]"){
  REQUIRE_THROWS_AS( Comm::TCP::Port(TEST_ADDRESS, TEST_PORT), Comm::TCP::ConnectionFailure );
}

TEST_CASE("tcp port connects to server", "[TCP_Port][.timing]"){
  Comm::TCP::SingleClientServer server(TEST_PORT);
  PAUSE
  Comm::TCP::Port port(TEST_ADDRESS, TEST_PORT);
  server.waitForConnection();
  REQUIRE( true );
}

TEST_CASE("tcp port can send data", "[TCP_Port][.timing]"){
  Comm::TCP::SingleClientServer server(TEST_PORT);
  PAUSE
  SECTION("client -> server"){
    Comm::TCP::Port sender(TEST_ADDRESS, TEST_PORT);
    Comm::TCP::Port receiver(server.waitForConnection());
    PAUSE
    REQUIRE_FALSE( receiver.hasData() );
    sender.push(std::string("test").c_str(), 4);
    PAUSE
    REQUIRE( receiver.hasData() );
  }
  SECTION("server -> client"){
    Comm::TCP::Port receiver(TEST_ADDRESS, TEST_PORT);
    Comm::TCP::Port sender(server.waitForConnection());
    PAUSE
    REQUIRE_FALSE( receiver.hasData() );
    sender.push(std::string("hello").c_str(), 4);
    PAUSE
    REQUIRE( receiver.hasData() );
  }
}

TEST_CASE("tcp port can read data", "[TCP_Port][.timing]"){
  Comm::TCP::SingleClientServer server(TEST_PORT);
  PAUSE
  SECTION("client -> server"){
    Comm::TCP::Port sender(TEST_ADDRESS, TEST_PORT);
    Comm::TCP::Port receiver(server.waitForConnection());
    sender.push(std::string("test").c_str(), 4);
    PAUSE
    char received[4];
    receiver.poll(received, 4);
    REQUIRE( std::string(received, 4) == std::string("test") );
  }
  SECTION("server -> client"){
    Comm::TCP::Port receiver(TEST_ADDRESS, TEST_PORT);
    Comm::TCP::Port sender(server.waitForConnection());
    sender.push(std::string("hello").c_str(), 5);
    PAUSE
    char received[5];
    receiver.poll(received, 5);
    REQUIRE( std::string(received, 5) == std::string("hello") );
  }
}

File: utils/javascript/src/comm/tcp/port.spec.js
const createPort = require('./port.js')
const createMockServer = require('../../../mock').createServer
let portfinder = require('portfinder')

let testPort = 3001
const testAddress = 'localhost'

let server = 0;

beforeEach(done => {
  portfinder.getPort(function (err, port) {
    testPort = port
    console.log('Running on port: ', testPort)
    server = createMockServer(testAddress, testPort)
    done()
  })
})

afterEach(() => {
  server.destroy()
  server = 0
})

test('tcp port can connect to server', done => {
  createPort(testAddress, testPort).then(port => {
    expect(port).toBeDefined()
    done()
  })
})

test('tcp port fails on connect if no server', done => {
  server.destroy(() => {
    createPort(testAddress, testPort).then(() => {
      throw 'TCP port should not connect.'
    }).catch(() => {
      done()
    })
  })
})

test('tcp port can send data', done => {
  server.connection(client => {
    client.on('data', data => {
      expect(data.toString()).toBe('test test test')
      done()
    })
  })
  createPort(testAddress, testPort).then(port => {
    port.push('test test test')
  })
})

test('tcp port should have data', done => {
  server.connection(client => {
    client.write('some data')
  })
  createPort(testAddress, testPort).then(port => {
    const checkData = () => {
      expect(port.hasData()).toBeTruthy()
      done()
    }
    setTimeout(checkData, 10)
  })
})

test('tcp port can read data', done => {
  server.connection(client => {
    client.write('hello world')
  })
  createPort(testAddress, testPort).then(port => {
    const checkData = () => {
      expect(port.poll()).toBe('hello world')
      expect(port.hasData()).toBeFalsy()
      done()
    }
    setTimeout(checkData, 10)
  })
})
